src/ClientHandle.cpp

```diff
@@ -773,31 +773,64 @@ void cClientHandle::HandleEnchantItem(UInt8 a_WindowID, UInt8 a_Enchantment)
 	)
 	{
+		LOGD("cClientHandle received a non valid enchanting window.");
 		return;
 	}
 
 	cEnchantingWindow * Window = static_cast<cEnchantingWindow *>(m_Player->GetWindow());
-	cItem Item = *Window->m_SlotArea->GetSlot(0, *m_Player);  // Make a copy of the item
 	short BaseEnchantmentLevel = Window->GetPropertyValue(a_Enchantment);
+	// Retrieve the enchanted item corresponding to our chosen option (top, middle, bottom)
+	cItem EnchantedItem = Window->m_SlotArea->GetEnchantedOption(a_Enchantment);
 
-	if (Item.EnchantByXPLevels(BaseEnchantmentLevel))
+	const bool IsCreative = m_Player->IsGameModeCreative();
+	const auto XpRequired = m_Player->XpForLevel(BaseEnchantmentLevel);
+	cItem LapisStack = *Window->m_SlotArea->GetSlot(1, *m_Player);  // Make a copy of the lapis stack
+	const auto LapisRequired = a_Enchantment + 1;
+
+	// Only allow enchantment if the player is creative or has sufficient levels and lapis to enchant
+	if (IsCreative || ((m_Player->GetCurrentXp() >= XpRequired) && LapisStack.GetCount() >= LapisRequired))
 	{
-		if (m_Player->IsGameModeCreative() || m_Player->DeltaExperience(-m_Player->XpForLevel(BaseEnchantmentLevel)) >= 0)
+		// If they weren't creative then subtract xp and lapis
+		if (!IsCreative)
 		{
-			Window->m_SlotArea->SetSlot(0, *m_Player, Item);
-			Window->SendSlot(*m_Player, Window->m_SlotArea, 0);
-			Window->BroadcastWholeWindow();
+			/** We need to reduce the player's level by the amount of lapis required
+			However we need to keep the resulting percentage filled the same */
+			// Get current level and fill percent
+			const auto CurrentLevel = m_Player->GetXpLevel();
+			const auto TargetLevel = CurrentLevel - LapisRequired;
+			const auto CurrentFillPercent = m_Player->GetXpPercentage();
+
+			// Remove current xp and add amount for target level
+			const auto DeltaForLevel = -m_Player->GetCurrentXp() + m_Player->XpForLevel(TargetLevel);
+			// Add correct amount to get the same fill percent
+			const auto DeltaForPercent = CurrentFillPercent*(m_Player->XpForLevel(TargetLevel + 1) - m_Player->XpForLevel(TargetLevel));
+			m_Player->DeltaExperience(DeltaForLevel + DeltaForPercent);
+
+			// Now reduce the lapis in our stack and send it back
+			LapisStack.AddCount(-LapisRequired);
+			Window->m_SlotArea->SetSlot(1, *m_Player, LapisStack);
+		}
+		// Set the item slot to our previously generated enchanted item
+		Window->m_SlotArea->SetSlot(0, *m_Player, EnchantedItem);
+		Window->BroadcastWholeWindow();
 
-			Window->SetProperty(0, 0, *m_Player);
-			Window->SetProperty(1, 0, *m_Player);
-			Window->SetProperty(2, 0, *m_Player);
+		// Reset window properties
+		for (short i=0; i<10; i++)
+		{
+			Window->SetProperty(i, 0, *m_Player);
 		}
+
+		// Permute the player's enchantment seed, this should probably be more fancy
+		auto EnchantmentSeed = m_Player->GetEnchantmentSeed();
+		MTRand Random(EnchantmentSeed);
+		m_Player->SetEnchantmentSeed(Random.RandInt<unsigned int>());
```

Comment:
Do we want to make this any more fancy, i.e. combine with the the exact item chosen to enchant or something? Might not matter.

Reply:
I personally think we don't need to. Random is random 😅

Reply:
Yea they shouldn't be able to calculate what enchantments they get next without their enchantment seed anyway.